nbconvert/preprocessors/execute.py

```diff
@@ -295,40 +240,44 @@ def setup_preprocessor(self, nb, resources, km=None):
         resources : dictionary
             Additional resources used in the conversion process.
         """
+        # Nbconvert can be installed without these, so do the imports only
+        # when the preprocessor is being used.
+        from jupyter_kernel_mgmt.discovery import KernelFinder
+        from jupyter_kernel_mgmt.client import BlockingKernelClient
+
         path = resources.get('metadata', {}).get('path', '') or None
         self.nb = nb
         # clear display_id map
         self._display_id_map = {}
 
-        if km is None:
-            self.km, self.kc = self.start_new_kernel(cwd=path)
-            try:
-                # Yielding unbound args for more easier understanding and downstream consumption
-                yield nb, self.km, self.kc
-            finally:
-                self.kc.stop_channels()
-                self.km.shutdown_kernel(now=self.shutdown_kernel == 'immediate')
-
-                for attr in ['nb', 'km', 'kc']:
-                    delattr(self, attr)
+        kf = KernelFinder.from_entrypoints()
+        if self.kernel_name:
+            kernel_name = self.kernel_name
+            if '/' not in kernel_name:
+                kernel_name = 'spec/' + kernel_name
         else:
-            self.km = km
-            if not km.has_kernel:
-                km.start_kernel(extra_arguments=self.extra_arguments, **kwargs)
-            self.kc = km.client()
-
-            self.kc.start_channels()
-            try:
-                self.kc.wait_for_ready(timeout=self.startup_timeout)
-            except RuntimeError:
-                self.kc.stop_channels()
-                raise
-            self.kc.allow_stdin = False
             try:
-                yield nb, self.km, self.kc
-            finally:
-                for attr in ['nb', 'km', 'kc']:
-                    delattr(self, attr)
+                kernel_name = 'spec/' + nb.metadata.get('kernelspec', {})[
```

Comment:
This will produce a duplicated `'spec/'` prefix when the existing notebook file has been persisted using the new kernel providers.  I modified this to the following when investigating how async support in jupyter_kernel_mgmt affects existing clients ([per this comment](https://github.com/takluyver/jupyter_kernel_mgmt/pull/23#issuecomment-525370746)).

```python
        # Check for unset kernel_name first and attempt to pull from metadata.  Only then
        # should we check for provider id prefixes, otherwise we could end up with a double prefix.
        kernel_name = self.kernel_name
        if not kernel_name:
            try:
                kernel_name = nb.metadata.get('kernelspec', {})['name']
            except KeyError:
                kernel_name = 'pyimport/kernel'

        # Ensure kernel_name is of the new form in case of older metadata
        if '/' not in kernel_name:
            kernel_name = 'spec/' + kernel_name
```